Sprint 2

alexvasoff opened the pull request:
README plz

alexvasoff committed:
feat: Add and configure ESLint

alexvasoff committed:
feat: Run ESLint --fix

alexvasoff committed:
feat: Add TypeScript and [ ESLint + TS = <3 ]

alexvasoff committed:
feat: Add StyleLint

alexvasoff committed:
feat: Add .editorconfig

alexvasoff committed:
feat: Off 'prefer-default-export' eslint rule

alexvasoff committed:
feat: Specify eslint config

alexvasoff committed:
feat: Add base BaseBlock (without id system)

alexvasoff committed:
feat: Add EventBus util

alexvasoff committed:
feat: Add renderDOM util

alexvasoff committed:
feat: Specify tsconfig.json

alexvasoff committed:
feat: Add uuid package

alexvasoff committed:
feat: Fixing dependency versions

alexvasoff committed:
feat: Specify tsconfig ruels

alexvasoff committed:
feat: Add HTTPTransport class

alexvasoff committed:
feat: Add _id property in baseBlock class

alexvasoff committed:
feat: Specify eslint

alexvasoff committed:
feat: add ButtonBlock

alexvasoff committed:
feat: add Button types description

alexvasoff committed:
feat: add InputBlock

alexvasoff committed:
feat: bad try

alexvasoff committed:
feat: baseBlock with children handle

alexvasoff committed:
feat: bad try

alexvasoff committed:
feat: Specify tsconfig

alexvasoff committed:
feat: Update README.md

agadoev commented on src/components/button/types.d.ts:
```diff
@@ -0,0 +1,5 @@
+export type Props = {
+  name: string;
+  text: string;
+  type?: string;
```
Полагаю, что тип у кнопки может принимать ограниченое количество значений, и если положить в него любую строку, то это будет ошибка. Предлагаю в таком случае воспользоваться discriminated union
```suggestion
  type?: 'button' | 'submit' и тд;
```

agadoev commented on README.md:
```diff
@@ -31,4 +31,25 @@
+
+Как грамотно применить добавленные механизмы, к сожалению, не успел разобраться.
+В задании с логированием данных форм заметил, что у меня блоки не были оформлены в тэгах <form>.
+Предпологаю, что надо все формы вообще выносить в отдельные компоненты (в условную папку forms) и импортировать в страницу (LoginPage <- LoginForm).
```
Это хорошая идея

agadoev commented on README.md:
```diff
@@ -31,4 +31,25 @@
+Как грамотно применить добавленные механизмы, к сожалению, не успел разобраться.
+В задании с логированием данных форм заметил, что у меня блоки не были оформлены в тэгах <form>.
+Предпологаю, что надо все формы вообще выносить в отдельные компоненты (в условную папку forms) и импортировать в страницу (LoginPage <- LoginForm).
+Скорее всего, добавил бы класс-обертку (может на основе baseBlock) для реализации логики логирования.
```
То, о чем ты говоришь, называет High Order Components. Это хороший вариант как прикрутить переиспользуемую логику в компонентам. Можно сделать функцию, например, ```withValidation```, которая будет принимать компонент, и возвращать другой компонент, который уже содержит логику валидации.

Есть путь проще - написать множество небольших функций-валидаторов, каждая из которых принимала бы строку, валидировала бы ее, и возвращала бы boolean, при создании input'a прокидывать эти функции в него через пропсы и в input'е уже валидировать на onblur/onfocus.

agadoev commented on src/pages/login/index.js:
```diff
@@ -1,15 +1,56 @@
+  name: 'register',
+  text: 'Регистрация',
+  type: 'text',
+}
```
Не нужно держать закомментированный код в репозитории, потому что он загрязняет кодовую базу

agadoev commented on src/utils/eventBus/index.ts:
```diff
@@ -0,0 +1,28 @@
+class EventBus {
+  listeners: {} = {};
```
Неплохо было бы написать типы для листенеров
```suggestion
  listeners: Record<string, (...args: unknown[]) => void>;
```

agadoev commented on src/utils/httpTransport/index.ts:
```diff
@@ -0,0 +1,72 @@
+
+  // Здесь достаточно и [object Object] для объекта
+  const keys = Object.keys(data);
+  return keys.reduce((result, key, index) => `${result}${key}=${data[key]}${index < keys.length - 1 ? '&' : ''}`, '?');
```
Выглядит очень сложно. Думаю, что можно изменить вот так
```suggestion
  return '?' + Object
    .entries(data)
    .map(([key, value]) => `${key}=${value}`)
    .join('&')
```

agadoev commented on src/utils/httpTransport/index.ts:
```diff
@@ -0,0 +1,72 @@
+const METHODS = {
```
кажется этот функционал относится к третьему спринту. Когда создаешь MR, старайся, чтобы в нем были только те изменения, которые относятся к задаче. Это упростит жизнь как ревьюеру, так и тебе

agadoev commented on tsconfig.json:
```diff
@@ -0,0 +1,14 @@
+    "module": "commonjs",
+    "target": "es2017",
+    "sourceMap": true,
+    "noImplicitAny": true,
```
Я бы добавил еще настройку ```noImplicitReturns```. Она запрещае неявные return'ы. Это сделает код более очевидным, а значит, с ним легче будет работать

agadoev reviewed (changes requested):
Привет, Алексей
Спасибо за работу, вижу компонентый подход. К сожалению, совсем без валидации принять работу не могу, попробуй все же ее реализовать, я написал в комментарии как это можно сделать. Из общих наблюдений - попробуй больше писать типов , потому что сейчас довольно мало когда покрыто TS'ом. При добавлении TS'a есть хороший способо как проверить, что TS используется хорошо. Попробуй сделать ошибку в сигнатуре/пропсах/строке, где угодно, и если TS ругнется, значит покрытие типами нормальное. Основная задача TS - предупреждать нас об ошибках на этапе написания кода.
Постарайся поправить те моменты, которые я указал и реализовать валидацию, затем двинемся дальше.

alexvasoff committed:
feat: Specify button type props via discriminated union

alexvasoff committed:
feat: remove unused commented code

alexvasoff committed:
feat: use index.ts instead index.js in Button component

alexvasoff committed:
fix: fix a misprint

alexvasoff committed:
feat: add 'noImplicitReturns' rule to tsconfig

alexvasoff committed:
feat: add types to listeners and another parts of EventBus

alexvasoff committed:
feat: Simplify statement

alexvasoff committed:
feat: Add RegExps and 'validate' function

alexvasoff committed:
feat: Disable rule of eslint

alexvasoff committed:
feat: Move InputFieldName type declaration from validator to types.d.ts of input component

alexvasoff committed:
feat: Switch type of file on login page

alexvasoff committed:
feat: Make style property not required in InputPropsType

alexvasoff committed:
feat: Add 'events' property to InputPropsType

alexvasoff committed:
feat: Add and use BaseProps type to baseComponent

alexvasoff committed:
feat: Rename function render to renderDOM

alexvasoff committed:
fix: Initial "listeners" field

alexvasoff committed:
feat: Specify element who get listeners

Это был интересный момент. Вроде не сложный, но забавный. Хороший "камень в огород" со стороны теории, подловила :)

alexvasoff committed:
fix: fix bug after renaming field

alexvasoff committed:
feat: move emit to main method

alexvasoff committed:
feat: remove the consequences of checks

alexvasoff committed:
feat: add empty line (git require, how I get it)

alexvasoff committed:
feat: create Input class with events, extends of BaseBlock

alexvasoff committed:
feat: change index.js to index.ts in input component

alexvasoff committed:
feat: update view logic

alexvasoff committed:
feat: Add bad tmp login props description

alexvasoff committed:
feat: Make 'events' field non required in type of input

alexvasoff committed:
feat: Birthday of first page on components and events <3

alexvasoff committed:
feat: Set default button type

alexvasoff committed:
feat: Specify type of InputFieldName

alexvasoff committed:
feat: Add bad tmp registration page props description

alexvasoff committed:
feat: Add passGuard to InputFieldName and regExp checker

alexvasoff committed:
feat: Add descriptions to regexps

Потом можно будет переработать объект regExp в структуру: 'trigger_name': {regexp, description}

alexvasoff committed:
feat: Make Registration page via class-components

alexvasoff committed:
feat: Add positive log to validation checker

alexvasoff committed:
feat: Rename index.js to index.ts in registration Page

alexvasoff committed:
feat: Update README.md

agadoev reviewed (approved):
Привет, Алексей
Спасибо за правки, вижу, что все поправлено. Супер, работа принимается, удачи в следующем спринте :)
